crates/cxx-qt-gen/src/preprocessor/self_inlining.rs

```diff
@@ -0,0 +1,172 @@
+// SPDX-FileCopyrightText: 2025 Klarälvdalens Datakonsult AB, a KDAB Group company <info@kdab.com>
+// SPDX-FileContributor: Ben Ford <ben.ford@kdab.com>
+//
+// SPDX-License-Identifier: MIT OR Apache-2.0
+
+use crate::parser::method::MethodFields;
+use crate::Parser;
+use proc_macro2::Ident;
+use quote::format_ident;
+use std::ops::DerefMut;
+use syn::spanned::Spanned;
+use syn::{Error, Result};
+
+/// Inline any `Self` types in the methods signatures with the Ident of a qobject passed in
+///
+/// If there are unresolved methods in the list, but inline is false, it will error,
+/// as the self inlining is only available if there is exactly one `QObject` in the block,
+/// and this indicates that no inlining can be done, but some `Self` types were present.
+pub fn try_inline_self_invokables(
+    inline: bool,
+    type_to_inline: &Option<Ident>,
+    invokables: &mut [impl DerefMut<Target = MethodFields>],
+) -> syn::Result<()> {
+    for method in invokables.iter_mut() {
+        if method.self_unresolved {
+            if inline {
+                if let Some(inline_type) = type_to_inline.clone() {
+                    method.qobject_ident = inline_type;
+                } else {
+                    return Err(Error::new(
+                        method.method.span(),
+                        "Expected a type to inline, no `qobject` typename was passed!",
+                    ));
+                }
+            } else {
+                return Err(Error::new(
+                    method.method.span(),
+                    "`Self` type can only be inferred if the extern block contains only one `qobject`.",
+                ));
+            }
+        }
+    }
+    Ok(())
+}
+
+/// For a given parser, attempt to inline the `Self` type used in any of the blocks with that blocks unique QObject
+pub fn qualify_self_types(parser: &mut Parser) -> Result<()> {
+    let data = &mut parser.cxx_qt_data;
+    for block in 0..data.qobjects.len() {
+        let methods = &mut data.methods[block];
+        let signals = &mut data.signals[block];
+        let inherited = &mut data.inherited_methods[block];
+        let qobjects = &data.qobjects[block];
+
+        let inline_self = qobjects.len() == 1;
```

Comment:
do we only ever need to care about QObjects? eg do we support non-qobject attributed types in extern C++Qt / RustQt blocks ?

and do we need to support normal `type` items in C++/Rust blocks if we start avoiding passthrough, this would be a separate change though?

Reply:
Right it's fine for now, but might need to change later depending what we do